duckduckgo/Android
# Move the app lifecycle observers to be plugins

Task/Issue URL: https://app.asana.com/0/414730916066338/1200284402862810/f
Tech Design URL: 
CC: 

**Description**:
The `DuckDuckGoAppliation` adds all application observers manually, like so:

```kotlin
ProcessLifecycleOwner.get().lifecycle.also {
  it.addObserver(this)
  it.addObserver(dataClearer)
  it.addObserver(appDaysUsedRecorder)
  it.addObserver(defaultBrowserObserver)
  it.addObserver(appEnjoymentLifecycleObserver)
  it.addObserver(dataClearerForegroundAppRestartPixel)
  it.addObserver(userStageStore)
  it.addObserver(pixelSender)
  it.addObserver(fireFireAnimationLoader)
  it.addObserver(tabsDbSanitizer)
  it.addObserver(webViewHttpAuthStore)
}
```

The main drawback is that the appliation has to depend on all the observer types, and every time we add/remove one observer we need to touch the applicaton.

We'd like to move that to use the plugin pattern, when the appliation will just get a set of `LifecycleObserver` plugins and add them. Liek so:

```kotlin
ProcessLifecycleOwner.get().lifecycle.also {
  it.addObserver(this)
  appLifecycleObserverPlugins.forEach { plugin ->
    it.addObserver(plugin)
  }
}
```

The `DuckDuckGoApplication` also manually initialised some other classes in lifecycle methods. I have also moved those to be a lifecycle observer. These are the classes:
* `appShortcutCreator.configureAppShortcuts()` was called in `onCreate` method. I moved that to the `AppShortcutCreatorLifecycleObserver` observer
* `appInstallStore` was initialised during `onCreate`. The `AppInstallStore` is also a lifecycle observer
* The `AppConfigurationSyncer` was initialised during `onCreate`. The `AppConfigurationSyncer` is also now a lifecycle observer

**Steps to test this PR**:
1. install/upgrade from this branch
2. For every observer listed above, add timbers in their methods annotated with `OnLifecycleEvent`
3. open the app
4. verify the methods are executed in the correct app lifecycle event

The DuckDuckApplication logs all registered observers, this should be the list:
```
com.duckduckgo.app.fire.AutomaticDataClearer
com.duckduckgo.app.global.shortcut.AppShortcutCreatorLifecycleObserver
com.duckduckgo.app.browser.httpauth.RealWebViewHttpAuthStore
com.duckduckgo.app.fire.LottieFireAnimationLoader
com.duckduckgo.app.global.install.AppInstallSharedPreferences
com.duckduckgo.app.global.rating.AppEnjoymentAppCreationObserver 
com.duckduckgo.app.browser.defaultbrowsing.DefaultBrowserObserver
com.duckduckgo.app.job.AppConfigurationSyncer
com.duckduckgo.app.statistics.api.RxPixelSender
com.duckduckgo.app.onboarding.store.AppUserStageStore
com.duckduckgo.app.usage.app.AppDaysUsedRecorder
com.duckduckgo.app.tabs.db.TabsDbSanitizer
com.duckduckgo.app.fire.DataClearerForegroundAppRestartPixel
```

1. instal/upgrade from this branch
2. ensure the core features of the app work as usual, bookmarks, fireproof sites, fire button, browsing


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
